src/peergos/server/tests/simulation/Simulator.java

```diff
@@ -0,0 +1,260 @@
+
+/**
+ * Run some I/O and then check the file-system is as expected.
+ */
+public class Simulator implements Runnable {
+    private static final Logger LOG = Logging.LOG();
+    private static final int MIN_FILE_LENGTH = 256;
+    private static final int MAX_FILE_LENGTH = Integer.MAX_VALUE;
+
+    enum Simulation {READ, WRITE, MKDIR}
+
+    private final int opCount, seed;
+    private final Random random;
+    private final double readProbability; // fraction of read ops
+    private final double mkdirProbability; // fraction of mkdir ops
+    private final double writeProbability; // fraction of mkdir ops
+    private final long meanFileLength;
+
+    private final FileSystem referenceFileSystem, testFileSystem;
+
+    private final Map<Path, List<String>> simulatedDirectoryToFiles = new HashMap<>();
+
+    long fileNameCounter;
+    long directoryNameCounter;
+
+    private String getNextName() {
+        return "" + fileNameCounter++;
+    }
+
+
+    private Path getRandomExistingFile() {
+
+        Path dir = getRandomExistingDirectory();
+        List<String> fileNames = simulatedDirectoryToFiles.get(dir);
+        int pos = random.nextInt(fileNames.size());
+        String fileName = fileNames.get(pos);
+        return dir.resolve(fileName);
+    }
+
+    private Path getRandomExistingDirectory() {
+        List<Path> directories = new ArrayList<>(simulatedDirectoryToFiles.keySet());
+
+        try {
+            int pos = random.nextInt(directories.size());
+            return directories.get(pos);
+        } catch (Exception ex) {
+            System.out.println();
+            throw ex;
+        }
+
+    }
+
+    private Path mkdir() {
+        String dirBaseName = getNextName();
+
+        Path path = getRandomExistingDirectory().resolve(dirBaseName);
+        simulatedDirectoryToFiles.putIfAbsent(path, new ArrayList<>());
+        return path;
+    }
+
+    private Path getAvailableFilePath() {
+        return getRandomExistingDirectory().resolve(getNextName());
+    }
+
+    private Simulation getNextSimulation() {
+        double d = random.nextDouble();
+        if (d < readProbability)
+            return Simulation.READ;
+        if (d < readProbability + writeProbability)
+            return Simulation.WRITE;
+        return Simulation.MKDIR;
+    }
+
+    private int getNextFileLength() {
+        double pos = random.nextGaussian();
+        int targetLength = (int) (pos * meanFileLength);
+        return Math.min(MAX_FILE_LENGTH,
+                Math.max(targetLength, MIN_FILE_LENGTH));
+    }
+
+    private byte[] getNextFileContents() {
+        byte[] bytes = new byte[getNextFileLength()];
+        random.nextBytes(bytes);
+        return bytes;
+    }
+
+    public Simulator(int opCount, int seed, double readProbability, double mkdirProbability, double writeProbability, long meanFileLength, FileSystem referenceFileSystem, FileSystem testFileSystem) {
+        if (readProbability + mkdirProbability > 1.0)
```

Comment:
You could check the sum of all three is within epsilon of 1.0

Reply:
this will change as we add more ops - will leave for now

Reply:
So it will still be incorrect then?

Reply:
All good